File: js/digital_signature_test.js
const jwt = require('jsonwebtoken');
require('dotenv').config();
const { expect } = require('chai');
const generate_digital_signature = require('./digital_signature');

const privateKey = process.env.PRIVATE.replace(/\\n/g, '\n');
const publicKey = process.env.PUBLIC.replace(/\\n/g, '\n');

const body = { name: 'John Doe', email: 'john@example.com' };


describe('generate_digital_signature', () => {
  it('should generate a valid digital signature', () => {
    const expectedSignature = jwt.sign(body, privateKey, { algorithm: 'RS256' });
    const actualSignature = generate_digital_signature(privateKey, body);
    expect(actualSignature).to.equal(expectedSignature);
  });

  it('should throw an error if digital key is invalid', () => {
    const invalidKey = 'invalid-digital-key';
    expect(() => generate_digital_signature(invalidKey, body)).to.throw();
  });

  it('should be able to verify the key', () => {
    const actualSignature = generate_digital_signature(privateKey, body);

    jwt.verify(actualSignature, publicKey, function(err, decoded) {
        expect(decoded.email).to.equal(body.email);
        expect(decoded.name).to.equal(body.name);
      });
  });

});

File: python/digital_signature_test.py
import unittest
import jwt
from digital_signature import generate_digital_signature
from dotenv import load_dotenv
import os

load_dotenv()  # Load the environment variables from the .env file


class TestGenerateDigitalSignature(unittest.TestCase):

    def setUp(self):
        self.private_key = os.environ.get("PRIVATE").replace("\\n", "\n")
        self.public_key = os.environ.get("PUBLIC").replace("\\n", "\n")
        self.body = {"user_id": 123}

    def test_generate_digital_signature(self):
        # Generate the digital signature using the function being tested
        signature = generate_digital_signature(self.private_key, self.body)
        # Verify that the signature was generated successfully
        self.assertIsNotNone(signature)

        # Verify that the signature can be decoded using the same digital key
        decoded = jwt.decode(signature, self.public_key, algorithms=["RS256"])
        self.assertDictEqual(decoded, self.body)


if __name__ == '__main__':
    unittest.main()
